File: node.js/letterPuzzle.ts
import * as fs from 'fs';
import { Rule, RuleSet, generateRuleList } from './affLogic';
import { fitsFilterCharCounts, generateWordList } from './dicLogic';

interface LetterBudget {
    chars: string;
    charCounts: number[];
    count: number;
};

function initLetterBudget(letters: string[], letterBudget: LetterBudget): void {
    letterBudget.chars = '';
    letterBudget.count = 0;

    for (const letter of letters) {
        letterBudget.chars += letter[0];
        letterBudget.count += letter.length;
        letterBudget.charCounts[letter[0]] = letter.length;
    };
};

function reduceBugdetByWord(word: string, budget: LetterBudget): LetterBudget {
    const result: LetterBudget = {
        chars: '',
        charCounts: [],
        count: 0
    };

    for (let i = 0; i < budget.chars.length; i++) {
        const filterChar = budget.chars[i];

        let filterCharCount = 0;
        for (let j = 0; j < word.length; j++) {
            if (word[j] === filterChar) {
                filterCharCount++;
            };
        };

        if (budget.charCounts[filterChar] > filterCharCount) {
            result.chars += filterChar;
            result.charCounts[filterChar] = budget.charCounts[filterChar] - filterCharCount;
        };

        result.count += budget.charCounts[filterChar] - filterCharCount;
    };

    return result;
};

function depthFirstSearch(wordOnlyOnce: Boolean, words: string[], startIndex: number, budget: LetterBudget, path: string, result: string[]): void {
    if (budget.count === 0) {
        result.push(path);

        if (result.length % 1000 === 0) {
            let duration = new Date().getTime() - start;
            console.log(`Count: ${result.length} (${(result.length / (duration / 1000))} / sec): ${result[result.length - 1]}`);
        };
    } else {
        for (let i = startIndex; i >= 0; i--) {
            const word = words[i];

            if (word.length > budget.count) {
                // words array is expected to be sorted by length.
                return;
            };

            if (fitsFilterCharCounts(word, budget.chars, budget.charCounts)) {
                if (wordOnlyOnce) {
                    depthFirstSearch(wordOnlyOnce, words, i - 1, reduceBugdetByWord(word, budget), path + ',' + word, result);
                } else {
                    depthFirstSearch(wordOnlyOnce, words, i, reduceBugdetByWord(word, budget), path + ',' + word, result);
                };
            };
        };
    };
};

function prepareWordList(words: string[], minWordLength: number): void {
    words.sort(function (a, b) {
        return b.length - a.length;
    });

    let i = words.length - 1;
    while (i >= 0 && words[i].length < minWordLength) {
        i--;
    };

    if (i < words.length - 1) {
        words.splice(i + 1);
    };
};

const lettersBudget: LetterBudget = {
    chars: '',
    charCounts: [],
    count: 0
};

//initLetterBudget(['eeeeeee', 'tt', 'l', 'iiiii', 'f', 'nnnn', 'ü', 'rr', 'u', 'h', 'c', 'ss', 'g', 'a', 'b', 'm'], lettersBudget);
initLetterBudget(['aa', 'e','hh','ll','o','i','r','d', 'nn', 't', 'u'], lettersBudget);
//initLetterBudget(['a', 'd', 'e', 'hh', 'ii', 'll', 'o', 'rr'], lettersBudget);
//initLetterBudget(['aa','hh','ll','o','i','r','d'], lettersBudget);
//initLetterBudget(['a', 'h', 'll', 'o'], lettersBudget);

const aff = fs.readFileSync('../hunspell/de_DE_frami_mod.aff');
const rules = generateRuleList(aff.toString().split('\r\n'));

const blackList = ['utrecht'];
const dic = fs.readFileSync('../hunspell/de_DE_frami.dic');
const words = generateWordList(dic.toString().split('\r\n'), rules, lettersBudget.chars, lettersBudget.charCounts, blackList);

prepareWordList(words, 2);

fs.writeFileSync('words.txt', words.join('\r\n'));

var start = new Date().getTime();

const result: string[] = [];
depthFirstSearch(true, words, words.length - 1, lettersBudget, '', result);

let duration = new Date().getTime() - start;
console.log(`Duration: ${duration /1000}sec Count: ${result.length} (${(result.length / (duration / 1000))} / sec): ${result[result.length - 1]}`);

fs.writeFileSync('combinations.txt', result.join('\r\n'));

File: cpp/letterPuzzle.cpp
#include "letterPuzzle.hpp"
#include "dicLogic.hpp"
#include <algorithm>
#include <map>
#include <iostream>

LetterBudget::LetterBudget(const std::string &word, LetterBudget &letterBudget)
{

    charCounts.reserve(letterBudget.chars.length());

    for (size_t i = 0; i < letterBudget.chars.length(); i++)
    {
        size_t filterCharCount = 0;

        for (const char &wordChar : word)
        {
            if (wordChar == letterBudget.chars.at(i))
            {
                filterCharCount += 1;
            }
        }

        if (letterBudget.charCounts.at(i) != filterCharCount)
        {
            chars += letterBudget.chars.at(i);
            charCounts.push_back(letterBudget.charCounts.at(i) - filterCharCount);
        }

        count += letterBudget.charCounts.at(i) - filterCharCount;
    }
}

LetterBudget::LetterBudget(const std::vector<std::string> &letters)
{
    for (const std::string &letter : letters)
    {
        count += letter.length();

        // Respect unicode.
        std::map<const char, size_t> chars;
        for (const char &chr : letter)
        {
            chars[chr] += 1;
        }

        for (const std::pair<const char, size_t> &charCount : chars)
        {
            this->chars += charCount.first;
            charCounts.push_back(charCount.second);
        }
    }
}

std::chrono::steady_clock::time_point start;

void depthFirstSearch(const bool &wordOnlyOnce, const std::vector<std::string> &words, const size_t &startIndex, LetterBudget &budget, const std::string &path, std::vector<std::string> &result)
{
    if (*budget.getCount() == 0)
    {
        result.push_back(path);

        if (result.size() % 1000 == 0)
        {
            int64_t duration = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::steady_clock::now() - start).count();
            std::cout << "Count: " << result.size() << " (" << result.size() / (duration / 1000000.0) << " / sec): " << result.at(result.size() - 1) << std::endl;
        }
    }
    else
    {
        for (size_t i = startIndex; i < words.size(); i++)
        {
            if (words.at(i).length() > *budget.getCount())
            {
                // words array is expected to be sorted by length.
                return;
            }

            if (fitsFilterCharCounts(words.at(i), *budget.getChars(), *budget.getCharCounts()))
            {
                LetterBudget letterBudget = LetterBudget(words.at(i), budget);
                if (wordOnlyOnce)
                {
                    depthFirstSearch(wordOnlyOnce, words, i + 1, letterBudget, path + ',' + words.at(i), result);
                }
                else
                {
                    depthFirstSearch(wordOnlyOnce, words, i, letterBudget, path + ',' + words.at(i), result);
                }
            }
        }
    }
}